   # VIKRAM SAINI

# Sonar Rock vs Mine Prediction Using Classification Techniques

1. Import Libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib as plt
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score,confusion_matrix,plot_confusion_matrix

2. Import The Dataset

In [2]:
Dataset = pd.read_csv("D:\Data Set\sonar data.csv ", header=None)
Dataset.head(3)

0       1       2       3       4       5       6       7       8   \
0  0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109   
1  0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
2  0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   

       9   ...      51      52      53      54      55      56      57  \
0  0.2111  ...  0.0027  0.0065  0.0159  0.0072  0.0167  0.0180  0.0084   
1  0.2872  ...  0.0084  0.0089  0.0048  0.0094  0.0191  0.0140  0.0049   
2  0.6194  ...  0.0232  0.0166  0.0095  0.0180  0.0244  0.0316  0.0164   

       58      59  60  
0  0.0090  0.0032   R  
1  0.0052  0.0044   R  
2  0.0095  0.0078   R  

[3 rows x 61 columns]

In [3]:
Dataset.shape

(208, 61)

3. Data Preprocessing

In [4]:
Dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 208 entries, 0 to 207
Data columns (total 61 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   0       208 non-null    float64
 1   1       208 non-null    float64
 2   2       208 non-null    float64
 3   3       208 non-null    float64
 4   4       208 non-null    float64
 5   5       208 non-null    float64
 6   6       208 non-null    float64
 7   7       208 non-null    float64
 8   8       208 non-null    float64
 9   9       208 non-null    float64
 10  10      208 non-null    float64
 11  11      208 non-null    float64
 12  12      208 non-null    float64
 13  13      208 non-null    float64
 14  14      208 non-null    float64
 15  15      208 non-null    float64
 16  16      208 non-null    float64
 17  17      208 non-null    float64
 18  18      208 non-null    float64
 19  19      208 non-null    float64
 20  20      208 non-null    float64
 21  21      208 non-null    float64
 22  22

Here one column is object and other are numerical values.

In [5]:
#Checking the Missing Values
Dataset.isnull().sum()

0     0
1     0
2     0
3     0
4     0
     ..
56    0
57    0
58    0
59    0
60    0
Length: 61, dtype: int64

Here is not any missing values.

In [6]:
Dataset.describe()

0           1           2           3           4           5   \
count  208.000000  208.000000  208.000000  208.000000  208.000000  208.000000   
mean     0.029164    0.038437    0.043832    0.053892    0.075202    0.104570   
std      0.022991    0.032960    0.038428    0.046528    0.055552    0.059105   
min      0.001500    0.000600    0.001500    0.005800    0.006700    0.010200   
25%      0.013350    0.016450    0.018950    0.024375    0.038050    0.067025   
50%      0.022800    0.030800    0.034300    0.044050    0.062500    0.092150   
75%      0.035550    0.047950    0.057950    0.064500    0.100275    0.134125   
max      0.137100    0.233900    0.305900    0.426400    0.401000    0.382300   

               6           7           8           9   ...          50  \
count  208.000000  208.000000  208.000000  208.000000  ...  208.000000   
mean     0.121747    0.134799    0.178003    0.208259  ...    0.016069   
std      0.061788    0.085152    0.118387    0.134416  ...    0.012008   
min      0.003300    0.005500    0.007500    0.011300  ...    0.000000   
25%      0.080900    0.080425    0.097025    0.111275  ...    0.008425   
50%      0.106950    0.112100    0.152250    0.182400  ...    0.013900   
75%      0.154000    0.169600    0.233425    0.268700  ...    0.020825   
max      0.372900    0.459000    0.682800    0.710600  ...    0.100400   

               51          52          53          54          55          56  \
count  208.000000  208.000000  208.000000  208.000000  208.000000  208.000000   
mean     0.013420    0.010709    0.010941    0.009290    0.008222    0.007820   
std      0.009634    0.007060    0.007301    0.007088    0.005736    0.005785   
min      0.000800    0.000500    0.001000    0.000600    0.000400    0.000300   
25%      0.007275    0.005075    0.005375    0.004150    0.004400    0.003700   
50%      0.011400    0.009550    0.009300    0.007500    0.006850    0.005950   
75%      0.016725    0.014900    0.014500    0.012100    0.010575    0.010425   
max      0.070900    0.039000    0.035200    0.044700    0.039400    0.035500   

               57          58          59  
count  208.000000  208.000000  208.000000  
mean     0.007949    0.007941    0.006507  
std      0.006470    0.006181    0.005031  
min      0.000300    0.000100    0.000600  
25%      0.003600    0.003675    0.003100  
50%      0.005800    0.006400    0.005300  
75%      0.010350    0.010325    0.008525  
max      0.044000    0.036400    0.043900  

[8 rows x 60 columns]

In [7]:
Dataset[60].value_counts()

M    111
R     97
Name: 60, dtype: int64

1. M - Mine
2. R - Rock

In [8]:
Dataset.groupby(60).mean()

0         1         2         3         4         5         6   \
60                                                                         
M   0.034989  0.045544  0.050720  0.064768  0.086715  0.111864  0.128359   
R   0.022498  0.030303  0.035951  0.041447  0.062028  0.096224  0.114180   

          7         8         9   ...        50        51        52        53  \
60                                ...                                           
M   0.149832  0.213492  0.251022  ...  0.019352  0.016014  0.011643  0.012185   
R   0.117596  0.137392  0.159325  ...  0.012311  0.010453  0.009640  0.009518   

          54        55        56        57        58        59  
60                                                              
M   0.009923  0.008914  0.007825  0.009060  0.008695  0.006930  
R   0.008567  0.007430  0.007814  0.006677  0.007078  0.006024  

[2 rows x 60 columns]

Using LabelEncoder change string in numerical values

In [9]:
label_encoder = LabelEncoder()
Dataset[60] = label_encoder.fit_transform(Dataset[60])
Dataset.head(3)

0       1       2       3       4       5       6       7       8   \
0  0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109   
1  0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
2  0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   

       9   ...      51      52      53      54      55      56      57  \
0  0.2111  ...  0.0027  0.0065  0.0159  0.0072  0.0167  0.0180  0.0084   
1  0.2872  ...  0.0084  0.0089  0.0048  0.0094  0.0191  0.0140  0.0049   
2  0.6194  ...  0.0232  0.0166  0.0095  0.0180  0.0244  0.0316  0.0164   

       58      59  60  
0  0.0090  0.0032   1  
1  0.0052  0.0044   1  
2  0.0095  0.0078   1  

[3 rows x 61 columns]

1. M = 0
2. R = 1

In [25]:
# Separating data and Lables
x = Dataset.iloc[:,:-1].values
Y = Dataset.iloc[:,-1]

In [26]:
# Split Dataset into training and test data set 
x_train,x_test,Y_train,Y_test = train_test_split(x,Y,test_size = 0.2,random_state=1)

In [28]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
x_train = sc.fit_transform(x_train)
x_test = sc.transform(x_test)

In [29]:
print(x.shape,x_train.shape,x_test.shape)

(208, 60) (166, 60) (42, 60)


# Logistic Regression

In [30]:
lr = LogisticRegression()
lr.fit(x_train,Y_train)

LogisticRegression()

Accuracy Score and Confusion Matrix for Training set

In [31]:
# Accuracy training data
Y_predict = lr.predict(x_train)
training_data_accuracy = accuracy_score(Y_predict,Y_train)
print(training_data_accuracy)

0.9457831325301205


In [32]:
Confusion_matrix_train = confusion_matrix(Y_predict,Y_train)
print(Confusion_matrix_train)

[[86  4]
 [ 5 71]]


In [33]:
#plot_confusion_matrix(model,Y_train,x_train)

Accuracy Score and Confusion Matrix for Test set

In [34]:
y_predict = lr.predict(x_test)
test_data_accuracy = accuracy_score(y_predict,Y_test)
print(test_data_accuracy)

0.7142857142857143


In [35]:
Confusion_matrix_test = confusion_matrix(y_predict,Y_test)
print(Confusion_matrix_test)

[[15  7]
 [ 5 15]]


# Support Vector Machine(SVM)

In [36]:
from sklearn.svm import SVC
svm = SVC(kernel='rbf',random_state=0)
svm.fit(x_train,Y_train)
Y_predict = svm.predict(x_train)

Accuracy Score and Confusion Matrix for Training set

In [37]:
Accuracy_Score_train = accuracy_score(Y_predict,Y_train)
print(Accuracy_Score_train)

0.9759036144578314


In [38]:
Confusion_matrix_train = confusion_matrix(Y_predict,Y_train)
print(Confusion_matrix_train)

[[88  1]
 [ 3 74]]


Accuracy Score and Confusion Matrix for Test set

In [39]:
y_predict = svm.predict(x_test)
Accuracy_Score_test = accuracy_score(y_predict,Y_test)
print(Accuracy_Score_test)

0.7380952380952381


In [40]:
Confusion_matrix_test = confusion_matrix(y_predict,Y_test)
print(Confusion_matrix_test)

[[16  7]
 [ 4 15]]


# K-nearest neighbors(KNN)

In [42]:
from sklearn.neighbors import KNeighborsClassifier
classifier = KNeighborsClassifier(n_neighbors = 5, metric = 'minkowski', p = 2)
classifier.fit(x_train, Y_train)

KNeighborsClassifier()

In [44]:
Y_predict = classifier.predict(x_train)
Accuracy_Score_train = accuracy_score(Y_predict,Y_train)
print(Accuracy_Score_train)

0.891566265060241


In [45]:
Confusion_matrix_train = confusion_matrix(Y_predict,Y_train)
print(Confusion_matrix_train)

[[84 11]
 [ 7 64]]


In [46]:
y_predict = svm.predict(x_test)
Accuracy_Score_test = accuracy_score(y_predict,Y_test)
print(Accuracy_Score_test)

0.7380952380952381


In [47]:
Confusion_matrix_test = confusion_matrix(y_predict,Y_test)
print(Confusion_matrix_test)

[[16  7]
 [ 4 15]]


# Training the Naive Bayes model on the Training set

In [48]:
from sklearn.naive_bayes import GaussianNB
n_b = GaussianNB()
n_b.fit(x_train, Y_train)

GaussianNB()

In [49]:
Y_predict = n_b.predict(x_train)
Accuracy_Score_train = accuracy_score(Y_predict,Y_train)
print(Accuracy_Score_train)

0.7349397590361446


In [51]:
Confusion_matrix_train = confusion_matrix(Y_predict,Y_train)
print(Confusion_matrix_train)

[[54  7]
 [37 68]]


In [53]:
# Making a Predictive System
input_data = (0.0388,0.0324,0.0688,0.0898,0.1267,0.1515,0.2134,0.2613,0.2832,0.2718,0.3645,0.3934,0.3843,0.4677,0.5364,0.4823,0.4835,0.5862,0.7579,0.6997,0.6918,0.8633,0.9107,0.9346,0.7884,0.8585,0.9261,0.7080,0.5779,0.5215,0.4505,0.3129,0.1448,0.1046,0.1820,0.1519,0.1017,0.1438,0.1986,0.2039,0.2778,0.2879,0.1331,0.1140,0.1310,0.1433,0.0624,0.0100,0.0098,0.0131,0.0152,0.0255,0.0071,0.0263,0.0079,0.0111,0.0107,0.0068,0.0097,0.0067)
#Changing the input_data to a numpy array
input_data_as_numpy_array = np.asarray(input_data)
# reshape the np array as we are predicting for one instance
input_data_reshaped = input_data_as_numpy_array.reshape(1,-1)

prediction = lr.predict(input_data_reshaped)
print(prediction)
if (prediction[0]=='R'):
    print("The object is a Rock")
else:
    print("The object is a Mine")


[0]
The object is a Mine


# -----------------------------------------------------------------------------------------------------------

In this dataset I use Classification techniques
1. Logistic Regression
2. Support Vector Machine(SVM)
3. K-NN
4. Naive Bayes

*Conclusion  -- SVM technique is more appropriate rather then other techniques.
* In SVM 98% accurate prediction for training set.